### ISSUE TITLE
Bug report: Hazelcast cluster discovery mode 'kubernetes' throws NoClassDefFoundError /FISH-6009

### ISSUE THREAD

**Issue by svendiedrichsen** [payara/Payara#5552]
# Description #
----------

When configuring payara server as documented [here](https://docs.payara.fish/community/docs/documentation/payara-server/hazelcast/discovery.html#_kubernetes_discovery_mode) to build a HZ data grid using the `kubernetes` discovery mode the server fails to start.

The same bug seems to appear when using the unsupported K8s DNS discovery as stated here: #5379 

## Expected Outcome ##

Payara server starts successfully with a Hazelcast cluster formed by using the kubernetes discovery mode as it used to do in version 5.2020.5.

## Current Outcome ##

Server fails to start with the following error:

```
java.lang.NoClassDefFoundError: com/hazelcast/cluster/Address
	at com.hazelcast.kubernetes.KubernetesApiEndpointResolver.createAddress(KubernetesApiEndpointResolver.java:101)
	at com.hazelcast.kubernetes.KubernetesApiEndpointResolver.addAddress(KubernetesApiEndpointResolver.java:83)
	at com.hazelcast.kubernetes.KubernetesApiEndpointResolver.getSimpleDiscoveryNodes(KubernetesApiEndpointResolver.java:76)
	at com.hazelcast.kubernetes.KubernetesApiEndpointResolver.resolve(KubernetesApiEndpointResolver.java:62)
	at com.hazelcast.kubernetes.HazelcastKubernetesDiscoveryStrategy.discoverNodes(HazelcastKubernetesDiscoveryStrategy.java:108)
	at com.hazelcast.spi.discovery.impl.DefaultDiscoveryService.discoverNodes(DefaultDiscoveryService.java:72)
	at com.hazelcast.internal.cluster.impl.DiscoveryJoiner.getPossibleAddresses(DiscoveryJoiner.java:71)
	at com.hazelcast.internal.cluster.impl.DiscoveryJoiner.getPossibleAddressesForInitialJoin(DiscoveryJoiner.java:60)
	at com.hazelcast.internal.cluster.impl.TcpIpJoiner.joinViaPossibleMembers(TcpIpJoiner.java:136)
	at com.hazelcast.internal.cluster.impl.TcpIpJoiner.doJoin(TcpIpJoiner.java:96)
	at com.hazelcast.internal.cluster.impl.AbstractJoiner.join(AbstractJoiner.java:137)
	at com.hazelcast.instance.impl.Node.join(Node.java:810)
	at com.hazelcast.instance.impl.Node.start(Node.java:473)
	at com.hazelcast.instance.impl.HazelcastInstanceImpl.<init>(HazelcastInstanceImpl.java:123)
	at com.hazelcast.instance.impl.HazelcastInstanceFactory.constructHazelcastInstance(HazelcastInstanceFactory.java:211)
	at com.hazelcast.instance.impl.HazelcastInstanceFactory.newHazelcastInstance(HazelcastInstanceFactory.java:190)
	at com.hazelcast.instance.impl.HazelcastInstanceFactory.newHazelcastInstance(HazelcastInstanceFactory.java:128)
	at com.hazelcast.core.Hazelcast.newHazelcastInstance(Hazelcast.java:57)
	at fish.payara.nucleus.hazelcast.HazelcastCore.bootstrapHazelcast(HazelcastCore.java:504)
	at fish.payara.nucleus.hazelcast.HazelcastCore.getInstance(HazelcastCore.java:239)
	at fish.payara.nucleus.config.ClusteredConfig.postConstruct(ClusteredConfig.java:100)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.glassfish.hk2.utilities.reflection.ReflectionHelper.invoke(ReflectionHelper.java:1268)
	at org.jvnet.hk2.internal.ClazzCreator.postConstructMe(ClazzCreator.java:309)
	at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:351)
	at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:463)
	at org.glassfish.hk2.runlevel.internal.AsyncRunLevelContext.findOrCreate(AsyncRunLevelContext.java:281)
	at org.glassfish.hk2.runlevel.RunLevelContext.findOrCreate(RunLevelContext.java:65)
	at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2102)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getService(ServiceLocatorImpl.java:667)
	at org.jvnet.hk2.internal.ThreeThirtyResolver.resolve(ThreeThirtyResolver.java:54)
	at org.jvnet.hk2.internal.ClazzCreator.resolve(ClazzCreator.java:188)
	at org.jvnet.hk2.internal.ClazzCreator.resolveAllDependencies(ClazzCreator.java:211)
	at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:334)
	at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:463)
	at org.glassfish.hk2.runlevel.internal.AsyncRunLevelContext.findOrCreate(AsyncRunLevelContext.java:281)
	at org.glassfish.hk2.runlevel.RunLevelContext.findOrCreate(RunLevelContext.java:65)
	at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2102)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:759)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:722)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getService(ServiceLocatorImpl.java:692)
	at com.sun.enterprise.v3.services.impl.ContainerMapper.<init>(ContainerMapper.java:113)
	at com.sun.enterprise.v3.services.impl.GlassfishNetworkListener.configureHttpProtocol(GlassfishNetworkListener.java:165)
	at org.glassfish.grizzly.config.GenericGrizzlyListener.configureSubProtocol(GenericGrizzlyListener.java:456)
	at org.glassfish.grizzly.config.GenericGrizzlyListener.configureSubProtocol(GenericGrizzlyListener.java:516)
	at org.glassfish.grizzly.config.GenericGrizzlyListener.configureProtocol(GenericGrizzlyListener.java:445)
	at org.glassfish.grizzly.config.GenericGrizzlyListener.configure(GenericGrizzlyListener.java:341)
	at com.sun.enterprise.v3.services.impl.GrizzlyProxy.initialize(GrizzlyProxy.java:126)
	at com.sun.enterprise.v3.services.impl.GrizzlyService.createNetworkProxy(GrizzlyService.java:673)
	at com.sun.enterprise.v3.services.impl.GrizzlyService.postConstruct(GrizzlyService.java:595)
	at org.jvnet.hk2.internal.ClazzCreator.postConstructMe(ClazzCreator.java:303)
	at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:351)
	at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:463)
	at org.glassfish.hk2.runlevel.internal.AsyncRunLevelContext.findOrCreate(AsyncRunLevelContext.java:281)
	at org.glassfish.hk2.runlevel.RunLevelContext.findOrCreate(RunLevelContext.java:65)
	at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2102)
	at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:93)
	at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:67)
	at org.glassfish.hk2.runlevel.internal.CurrentTaskFuture$QueueRunner.oneJob(CurrentTaskFuture.java:1213)
	at org.glassfish.hk2.runlevel.internal.CurrentTaskFuture$QueueRunner.run(CurrentTaskFuture.java:1144)
	at org.glassfish.hk2.runlevel.internal.CurrentTaskFuture$UpOneLevel.run(CurrentTaskFuture.java:762)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ClassNotFoundException: com.hazelcast.cluster.Address not found by com.hazelcast.kubernetes [377]
	at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1585)
	at org.apache.felix.framework.BundleWiringImpl.access$300(BundleWiringImpl.java:79)
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:1970)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 67 more
```
## Steps to reproduce ## 

Follow the documentation to configure Payara server to form a Hazelcast cluster using the `Kubernetes Discovery Mode`

https://docs.payara.fish/community/docs/documentation/payara-server/hazelcast/discovery.html#_kubernetes_discovery_mode

## Environment ##

- **Distribution**:  Server Full Profile 5.2021.10
- **JDK Version**: OpenJDK 11
- **Operating System**: Linux

**Comment by svendiedrichsen** [comment]
It may be a Hazelcast 4.x bug with the OSGI bundle meta information lacking the import of the required `com.hazelcast.cluster` package name.

**Comment by svendiedrichsen** [comment]
@fturizo Could someone from Payara confirm the bug and reach out to Hazelcast to kindly consider the fix? Seems to me that the Hazelcast integration is an important part of the Payara clustering efforts.

**Comment by fturizo** [comment]
Greetings @svendiedrichsen, sorry for the late reply on this issue, but we had other priorities to attend. After testing a simple scenario of using Payara Server's `kubernetes` discovery mode in the Data Grid, indeed there's an issue that prevents it from correctly starting, reporting the same stack trace that you supplied. It seems that this error was introduced on release 5.2021.1. 

We have raised the internal JIRA issue `FISH-6009` to fix this issue as soon as possible since this is a core common use case of the server and thus must work correctly.

**Comment by svendiedrichsen** [comment]
@fturizo Thank you. I have raised a bug with PR to fix it at the hazelcast-kubernetes project. Maybe it could be possible to reach out to Hazelcast for considering the fix and building a release of their lib.

**Comment by fturizo** [comment]
@svendiedrichsen, thanks for the reminder on the bug report + PR contribution that you've raised on the `hazelcast-kubernetes` project. I've passed over this info to our engineers so they take it into account.

**Comment by svendiedrichsen** [comment]
Bug fixed in version 5.2022.2. Thank you.
